pkg/services/sqlstore/migrations/ualert/tokenizer_test.go

```diff
@@ -0,0 +1,232 @@
+		text:  "${@instance1} is down",
+		token: Token{Variable: "@instance1"},
+		pos:   13,
+	}, {
+		name:  "two variables without spaces",
+		text:  "${variable1}${variable2}",
+		token: Token{Variable: "variable1"},
+		pos:   12,
+	}, {
+		name: "variable with newline",
+		text: "${instance\n} is down",
+		pos:  10,
+		err:  "unexpected whitespace",
+	}, {
+		name: "all '$' returns error",
+		text: "$$",
+		pos:  1,
+		err:  "expected '{', got '$'",
+	}, {
+		name: "two '$' before variable returns error",
+		text: "$${instance} is down",
+		pos:  1,
+		err:  "expected '{', got '$'",
+	}, {
+		name: "variable without braces returns error",
+		text: "$instance is down",
+		pos:  1,
+		err:  "expected '{', got 'i'",
+	}, {
+		name: "variable with two opening braces returns error",
+		text: "${{instance}",
+		pos:  2,
+		err:  "unexpected '{'",
+	}, {
+		name: "variable without closing brace returns error",
+		text: "${instance",
+		pos:  10,
+		err:  "expected '}', got 'e'",
+	}, {
+		name: "variable without closing and literal brace returns error",
+		text: "${instance is down",
+		pos:  18,
+		err:  "expected '}', got 'n'",
+	}, {
+		name: "variable with two closing braces returns error",
+		text: "${instance}} is down",
+		pos:  11,
+		err:  "unexpected '}'",
+	}, {
+		name: "variable with nested braces returns error",
+		text: "${instance{}} is down",
+		pos:  10,
+		err:  "unexpected '{'",
+	}, {
+		name: "variable with nested '$' returns error",
+		text: "${instance$} is down",
+		pos:  10,
+		err:  "unexpected '$'",
+	}, {
+		name: "variable with extra '}' returns error",
+		text: "${instance}} is down",
+		pos:  11,
+		err:  "unexpected '}'",
+	}}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			token, pos, err := tokenizeVariable([]rune(test.text))
+			if test.err != "" {
+				assert.EqualError(t, err, test.err)
+			}
+			assert.Equal(t, test.pos, pos)
+			assert.Equal(t, test.token, token)
+		})
+	}
+}
+
+func TestTokenizeTmpl(t *testing.T) {
+	tests := []struct {
+		name   string
+		tmpl   string
+		tokens []Token
+		err    string
+	}{{
+		name:   "simple template can be tokenized",
+		tmpl:   "${instance} is down",
+		tokens: []Token{{Variable: "instance"}, {Literal: " is down"}},
+	}, {
+		name: "complex template can be tokenized",
+		tmpl: "More than ${value} ${status_code} in the last 5 minutes",
+		tokens: []Token{
+			{Literal: "More than "},
+			{Variable: "value"},
+			{Literal: " "},
+			{Variable: "status_code"},
+			{Literal: " in the last 5 minutes"},
+		},
+	}, {
+		name:   "variables without spaces between can be tokenized",
+		tmpl:   "${value}${status_code}",
+		tokens: []Token{{Variable: "value"}, {Variable: "status_code"}},
+	}, {
+		name:   "variables without spaces between then literal can be tokenized",
+		tmpl:   "${value}${status_code} in the last 5 minutes",
+		tokens: []Token{{Variable: "value"}, {Variable: "status_code"}, {Literal: " in the last 5 minutes"}},
+	}}
```

Comment:
Here are the test cases I considered when trying out the above reparse changes:

```go
{
		name:   "",
		tmpl:   "Instance $${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "$"}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance $instance and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "$instance and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance ${{instance} and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "${{instance} and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance ${instance and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "${instance and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance ${instance}} and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Variable: "instance"}, {Literal: "} and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance ${instance{}} and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "${instance{}} and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "Instance ${instance$} and ${instance} is down",
		tokens: []Token{{Literal: "Instance "}, {Literal: "${instance"}, {Literal: "$} and "}, {Variable: "instance"}, {Literal: " is down"}},
	}, {
		name:   "",
		tmpl:   "${value}${status_code\n}${value} in the last 5 minutes",
		tokens: []Token{{Variable: "value"}, {Literal: "${status_code\n}"}, {Variable: "value"}, {Literal: " in the last 5 minutes"}},
	}, {
		name:   "",
		tmpl:   "${value}${status_code}}${value} in the last 5 minutes",
		tokens: []Token{{Variable: "value"}, {Variable: "status_code"}, {Literal: "}"}, {Variable: "value"}, {Literal: " in the last 5 minutes"}},
	}
```

or if you prefer, you can add them to `alert_rule_test.go` to test the feature outside the tokenizer implementation:

```go
func TestTemplateMigration(t *testing.T) {
	tt := []struct {
		name             string
		legacyMessage    string
		expectedTemplate string
	} {
		...
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
                        ...
			require.Equal(t, tc.expectedTemplate, ar.Annotations["message"])
		})
	}
}
```